[Merged by Bors] - Add regression tests for boot_node

mooori opened the pull request:
## Issue Addressed
Resolves #2602

## Proposed Changes

*Note: For a review it might help to look at the individual commits.*

### `boot_node`
Add support for the flags `dump-config` and `immediate-shutdown`. For `immediate-shutdown` the actual behavior could be described as `dump-config-and-exit`.

Both flags are handled in `boot_node::main`, which appears to be the simplest approach.

### `boot_node` regression tests
Added in `lighthouse/tests/boot_node.rs`.

### `CommandLineTestExec`
Factors out boilerplate related to CLI tests. It's used in the regression tests for `boot_node`, `beacon_node` and `validator_client`.

## Open TODO
Add tests for `boot_node` flags `enable-enr-auto-update` and `disable-packet-filter`. They end up in [`Discv5Config`](https://github.com/mooori/lighthouse/blob/9ed2cba6bc3e41f08207cb0eeaf9e4aee40d05dd/boot_node/src/config.rs#L29), which doesn't support serde (de)serialization.

I haven't found a workaround - guidance would be appreciated.

mooori committed:
add serde (de)serialization for BootNodeConfig

mooori committed:
add support for dump-config flag in boot_node

mooori committed:
support immediate-shutdown flag in boot_node

mooori committed:
merge updated upstream/unstable

mooori committed:
Add trait `CommandLineTestExec`

It aims to reduce boilerplate in regression tests for CLI flags.

mooori committed:
Add regression tests for `boot_node` CLI flags

mooori committed:
Merge remote-tracking branch 'upstream/unstable'

mooori committed:
Update use declarations

mooori committed:
Use `CommandLineTestExec` in `beacon_node` CLI tests

mooori committed:
Add `CommandLineTestExec::run_with_no_datadir`

Clarifying comments as well.

mooori committed:
Fix outdated comment

mooori committed:
Use `CommandLineTestExec` in `validator_client` CLI tests

mooori committed:
Merge remote-tracking branch 'upstream/unstable' into boot-node-regr-tests

mooori committed:
Remove `lighthouse/tests/common`

The `common` module was created as I had set `autotests = true` during
development of  the `boot_node` tests.

This commit makes `lighthouse/tests` work when compiled as single crate.

mooori committed:
Merge remote-tracking branch 'upstream/unstable' into boot-node-regr-tests

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/sigp/lighthouse?pullRequest=2749) <br/>All committers have signed the CLA.

paulhauner commented:
Perhaps @macladson would be interested in giving this a review?

macladson commented on lighthouse/tests/boot_node.rs:
```diff
@@ -0,0 +1,141 @@
+}
+
+#[test]
+fn boot_nodes_flag() {
```
I'd be tempted to check all the boot nodes here rather than just the extra ones. It'll help make sure that the hard-coded boot nodes are properly finding their way into the config.

mooori replied:
Done in 77e0be5

macladson commented on boot_node/src/config.rs:
```diff
@@ -2,26 +2,41 @@ use beacon_node::{get_data_dir, get_eth2_network_config, set_network_config};
+    // does not implement Serialize and Deserialize.
+    // Despite skip a default is needed to enable BootNodeConfig's
+    // Deserialization (CombinedKey has no default value).
+    #[serde(skip, default = "generate_local_key")]
```
Unfortunately, using `#[serde(skip, default = "generate_local_key")]` to circumvent the serde problem, comes with it's own set of issues.

The first issue that arises is that `generate_local_key()` uses an `unwrap`. Now in theory, this particular piece of code should never run outside of the `boot_node` tests, however having the possiblity of an unhandled panic in code which isn't explicitly marked as a test is generally something we want to avoid in Lighthouse.

The second issue is that now that the `BootNodeConfig` struct is serializable and deserializable, it's possible that in the future someone will attempt to use it as such (outside of tests). When deserializing `local_key`,  `default = "generate_local_key"` will then generate a new `CombinedKey` for it which could be misleading.

It could potentially be a bit messy, but a safer solution would be to define a new struct which contains just the (de)serializable values:
```rust
#[derive(Serialize, Deserialize)]
pub struct BootNodeConfigFile { 
	pub listen_socket: SocketAddr,
	// TODO: Generalise to multiaddr
	pub boot_nodes: Vec<Enr>, 
	pub local_enr: Enr, 
}
```

You can then instantiate this struct by pulling the values out of the defined `BootNodeConfig` using struct expansion. I haven't tested it, but it'll be something like:

```rust
let config: BootNodeConfig<T> = BootNodeConfig::try_from(matches)?;

let BootNodeConfig<T> {
	listen_socket,
	boot_nodes,
	local_enr,
	..
} = config;

let config_file = BootNodeConfigFile {
	listen_socket,
	boot_nodes,
	local_enr
};
```

Then you can pass `config_file` to the serializer as before.

Doing it this way also ensures the two structs are consistent since it'll start throwing syntax errors if the relevant fields ever change on one but not the other.

mooori replied:
Done in 482409e which introduced lint errors.
They are fixed in 37c3bf2

macladson commented on lighthouse/src/main.rs:
```diff
@@ -204,7 +204,13 @@ fn main() {
+            boot_node::run(
+                bootnode_matches,
+                eth_spec_id,
+                clap_utils::parse_optional::<PathBuf>(&matches, "dump-config")?,
```
This is probably fine as is. It *might* be worth just passing in `matches` instead, then parsing out the needed fields later in `boot_node/src/lib.rs`. It'll probably be cleaner and make it easier in the future if we ever need access to additional fields.

mooori replied:
Done in 6a3cecc.

macladson replied:
I believe `debug_info` is inside `matches` so we can probably remove it here and parse it inside of `boot_node` as well.

mooori replied:
I'm marking this as resolved to let us discuss it below.

macladson commented on lighthouse/tests/beacon_node.rs:
```diff
@@ -404,26 +320,28 @@ fn default_boot_nodes() {
     "enr:-LK4QKWrXTpV9T78hNG6s8AM6IO4XH9kFT91uZtFg1GcsJ6dKovDOr1jtAAFPnS2lvNltkOGA9k29BUN7lFh_sjuc9QBh2F0dG5ldHOIAAAAAAAAAACEZXRoMpC1MD8qAAAAAP__________gmlkgnY0gmlwhANAdd-Jc2VjcDI1NmsxoQLQa6ai7y9PMN5hpLe5HmiJSlYzMuzP7ZhwRiwHvqNXdoN0Y3CCI4yDdWRwgiOM"
```
Would you mind taking the opportunity to fix the (lack of) indent [here](https://github.com/sigp/lighthouse/blob/9ed2cba6bc3e41f08207cb0eeaf9e4aee40d05dd/lighthouse/tests/beacon_node.rs#L305).

mooori replied:
Done in 17a28d6.

macladson reviewed (changes requested):
This is awesome, thanks so much for doing this!
Overall it's very clean and I love the way you handled the refactor.
I do have a couple of suggestions for you, but the only major one is around handling the (de)serialization of the `BootNodeConfig`.

mooori commented:
Thank you @macladson for your review and suggestions!

I'm going to tackle these points :)

mooori committed:
Merge remote-tracking branch 'upstream/unstable' into boot-node-regr-tests

mooori committed:
Refactor `BootNodeConfig` (de)serialization

The (de)serialization of `BootNodeConfig` introduced ambiguities and
the possibility of an unhandled panic.

To circumvent these issues, (de)serialization is handled via
`BootNodeConfigSerialization`.

mooori committed:
Clean up `boot_node::run` parameters

mooori committed:
Extend test `boot_node::boot_nodes_flag`

Ensure hardcoded boot nodes end up in the config.

mooori committed:
Fix missing indent

mooori committed:
Merge remote-tracking branch 'upstream/unstable' into boot-node-regr-tests

mooori committed:
Fix lint errors

Related to (de)serialization of `BootNodeConfig`.

mooori commented:
Just pushed the changes and from my side this would be ready for review.

macladson commented on lighthouse/src/main.rs:
```diff
@@ -204,7 +204,7 @@ fn main() {
                 .into();
 
-            boot_node::run(bootnode_matches, eth_spec_id, debug_info);
+            boot_node::run(&matches, bootnode_matches, eth_spec_id, debug_info);
```
As per earlier comment, I believe `debug_info` is inside `matches` so we can probably remove it here and parse it inside of `boot_node` as well.

mooori replied:
While working on the changes I was about to do this, but then realized it moves the possibility of [this](https://github.com/mooori/lighthouse/blob/37c3bf2cc44db2d8fc40b5abe1fbe889482fb128/lighthouse/src/main.rs#L204) unhandled panic from `main` to `boot_node::run`.

If that's okay, I'm happy to move parsing of `debug-level` to `boot_node::run`.

Or we change `boot_node::run` to return a `Result` and then keep the panic in `main`?

macladson replied:
Ah yes, that's fair. I don't *think* there would be any problem with doing that, but there's probably no need to rock the boat on this one right now.

macladson reviewed (changes requested):
Awesome! Just one more thing but this looks good to me otherwise.

macladson commented:
Could I get @paulhauner to run CI on this please?

paulhauner commented on boot_node/src/config.rs:
```diff
@@ -130,3 +131,26 @@ impl<T: EthSpec> TryFrom<&ArgMatches<'_>> for BootNodeConfig<T> {
+impl BootNodeConfigSerialization {
+    /// Returns a `BootNodeConfigSerialization` obtained from copying resp. cloning the
+    /// relevant fields of `config`
+    pub fn from_config_ref<T: EthSpec>(config: &BootNodeConfig<T>) -> Self {
```
If we adopt the following pattern, we can force a compile time error if a field is added to `BeaconNodeConfig` without also considering if it should be added to `BeaconNodeConfigSerialization`:

```rust
    pub fn from_config_ref<T: EthSpec>(config: &BootNodeConfig<T>) -> Self {
        let BootNodeConfig {
            listen_socket,
            boot_nodes,
            local_enr,
            local_key: _,
            discv5_config: _,
            phantom: _,
        } = config;

        BootNodeConfigSerialization {
            listen_socket: *listen_socket,
            boot_nodes: boot_nodes.clone(),
            local_enr: local_enr.clone(),
        }
    }
```

It's a little more verbose, but I think it would be nice to force the future developer to consider the relationship between these structs.

mooori replied:
Neat! Done in 71013e7.

paulhauner reviewed (changes requested):
This is great and almost there! I just have one minor request please :pray:

mooori committed:
Merge remote-tracking branch 'upstream/unstable' into boot-node-regr-tests

mooori committed:
Resolve merge conflict

mooori committed:
Refactor `BootNodeConfigSerialization::from_config_ref`

Add pattern to enforce handling all fields of `BootNodeConfig`.

paulhauner reviewed (approved):
Great work, thank you!

paulhauner commented:
bors r+
